File: ClimbingStairs.py
"""
You are climbing a stair case. It takes n steps to reach to the top.

Each time you can either climb 1 or 2 steps. In how many distinct ways can you climb to the top?
"""


class Solution:
    # @param {integer} n
    # @return {integer}
    def climbStairs(self, n):
        if 0 <= n <= 2:
            return n
        steps = [1, 2]
        for _ in xrange(n - 2):
            steps.append(steps[-2] + steps[-1])
        return steps[-1]


if __name__ == '__main__':
    assert Solution().climbStairs(3) == 3

File: Leetcode-Java/ClimbingStairs.java
/*
You are climbing a stair case. It takes n steps to reach to the top.

Each time you can either climb 1 or 2 steps. In how many distinct ways can you climb to the top?
*/

import java.util.*;

public class ClimbingStairs {
	public static int climbingStairs(int n) {
		List<Integer> l = new ArrayList<Integer>();
		l.add(0);
		l.add(1);
		l.add(2);
		if (n <= 2) {
			return l.get(n);
		}
		for (int i = 3; i <= n; i++) {
			l.add(l.get(i-2) + l.get(i-1));
		}
		return l.get(n);
	}

	public static void main(String[] args) {
		System.out.println(climbingStairs(1));
		System.out.println(climbingStairs(2));
		System.out.println(climbingStairs(3));
		System.out.println(climbingStairs(4));
		System.out.println(climbingStairs(10));
	}
}
